jib-core/src/main/java/com/google/cloud/tools/jib/api/ImageReference.java

```diff
@@ -346,6 +387,10 @@ public String toString() {
    * @return the image reference in Docker-readable format, without hiding the tag
    */
   public String toStringWithTag() {
+    // Insert tag before digest
+    if (!Strings.isNullOrEmpty(digest)) {
+      return toString();
```

Comment:
I think the purpose of this method is to ensure `:latest` is added. I don't think `toString()` will add `:latest`.

Reply:
I'll have to check how it's used in the rest of the codebase, but this is used for the offline mode cache directory structure. So if we add "latest" here, people may need to repopulate their cache.

Reply:
Ugh, the offline manifest caching! So we want that to be more like `toStringWithTagOrDigest()` where a digest takes precedence, right? The naming is a bit unfortunate. `toStringWithTag` doesn't make sense and is confusing. Should we just create a new method? I wish we could create an internal-only method for the offline caching.

Reply:
Whether or not we deprecate this method, I think this method should be updated to fulfill the documented contract so that it returns `registry/repository:latest@sha256...` when both a tag and a digest are given. Given the name `toStringWithTag()`, returning `registry/repository@sha256...` doesn't look right to me.

I'm struggling with the name for the new method too. I don't think there's a good name to cover both tag and digest. This [grammar description](https://github.com/docker/distribution/blob/742aab907b54a367e1ac7033fb9fe73b0e7344f5/reference/reference.go#L6-L24) may be worth checking, although I don't know where the description came from. At least `toStringWithQualifier` sounds better than "identifier".

Anyways, maybe we should mark the new method as "internal"–the purpose is too specific to our use-case. @loosebazooka had some discussions around how we could add internal methods to our public API, but we couldn't find a good solution. The best I could do was to [document it](https://github.com/GoogleContainerTools/jib/pull/2353/files#diff-d390f64e06290a875d4e6a149d0f18ccR580).

Reply:
>Given the name toStringWithTag(), returning registry/repository@sha256... doesn't look right to me.

The problem is that this was what the old behavior was, regardless of whether or not the naming was correct, since we were treating digests as tags. Is it more important to preserve old behavior in this case or change the behavior to better match the name?

Reply:
Ok, I think given jib-core still isn't GA, we can just rename the method now instead of having to deprecate.